### ISSUE TITLE
Maximum call stack size exceeded when using circular dependencies inside a scoped container

### ISSUE THREAD

**Issue by lenartbezek** [typestack/typedi#112]
When trying to get services inside a scoped container that reference each other,
we get `Maximum call stack size exceeded error`.

Code required to reproduce:
```ts
import "reflect-metadata";
import Container, { Service, Inject } from "typedi";

@Service()
class A {
  @Inject(() => B)
  b: any;
}

@Service()
class B {
  @Inject(() => A)
  a: any;
}

// Works
console.log({
  a: Container.get(A),
  b: Container.get(B),
});

const instance = Container.of("some-instance");
// Throws Maximum call stack size exceeded error
console.log({
  a: instance.get(A),
  b: instance.get(B),
});
```

Output when running with `ts-node`:
```
PS D:\typedi-test> ts-node --project .\tsconfig.json index.ts
{ a: A { b: B { a: [Circular] } }, b: B { a: A { b: [Circular] } } }

D:\typedi-test\src\ContainerInstance.ts:245
        return this.services.find(service => {
                             ^
RangeError: Maximum call stack size exceeded
    at D:\typedi-test\src\ContainerInstance.ts:245:35
    at Array.find (<anonymous>)
    at ContainerInstance.findService (D:\typedi-test\src\ContainerInstance.ts:245:30)
    at ContainerInstance.get (D:\typedi-test\src\ContainerInstance.ts:102:39)
    at Object.value (D:\typedi-test\src\decorators\Inject.ts:48:42)
    at D:\typedi-test\src\ContainerInstance.ts:381:54
    at Array.forEach (<anonymous>)
    at ContainerInstance.applyPropertyHandlers (D:\typedi-test\src\ContainerInstance.ts:376:28)
    at ContainerInstance.getServiceValue (D:\typedi-test\src\ContainerInstance.ts:343:18)
    at ContainerInstance.get (D:\typedi-test\src\ContainerInstance.ts:114:32)
```
No idea why source paths are wrong there ...

Error when running compiled code:
```
PS D:\typedi-test> node .\dist\index.js
{ a: A { b: B { a: [Circular] } }, b: B { a: A { b: [Circular] } } }
D:\typedi-test\node_modules\reflect-metadata\Reflect.js:562
        function GetOrCreateMetadataMap(O, P, Create) {
                                       ^

RangeError: Maximum call stack size exceeded
    at GetOrCreateMetadataMap (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:562:40)
    at OrdinaryHasOwnMetadata (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:593:31)
    at OrdinaryGetMetadata (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:601:26)
    at OrdinaryGetMetadata (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:606:24)
    at OrdinaryGetMetadata (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:606:24)
    at Object.getMetadata (D:\typedi-test\node_modules\reflect-metadata\Reflect.js:357:20)
    at ContainerInstance.getServiceValue (D:\typedi-test\node_modules\typedi\ContainerInstance.js:181:75)
    at ContainerInstance.get (D:\typedi-test\node_modules\typedi\ContainerInstance.js:47:30)
    at Object.value (D:\typedi-test\node_modules\typedi\decorators\Inject.js:30:42)
    at D:\typedi-test\node_modules\typedi\ContainerInstance.js:246:54
```

**Comment by lenartbezek** [comment]
Apparently this can be fixed by copying services to a scoped container _at time of creation_, instead of relying on them being cloned from global container at `get`.

Solution would look something like this:
```ts
// in Container.of when creating a new container instance
container = new ContainerInstance(instanceId);
container.services.push(...this.globalInstance.services.map(s => ({
    ...s,
    value: s.global ? s.value : undefined
})));
```

I need opinion on whether this is an appropriate solution.

**Comment by lenartbezek** [comment]
https://github.com/typestack/typedi/compare/master...lenartbezek:master

**Comment by woubuc** [comment]
I'm running into this issue atm. Seems like there is a fix, when will the package be updated on npm?

**Comment by kiurchv** [comment]
@pleerock @lenartbezek any chance to release fix for this issue?

**Comment by NoNameProvided** [comment]
Hi @lenartbezek! Thanks for the report! Can you open a PR for this please?

**Comment by NoNameProvided** [comment]
Hi! 

Thanks for the PR! I was careless and didn't see you target the `master` branch. Can you create a PR against the `develop` branch? I have tried to do it but it seems your test-case is failing on develop. (needs to be adapted for Jest)

```ts
import "reflect-metadata";

import {Container} from "../../../src/Container";
import {Service} from "../../../src/decorators/Service";
import {Inject} from "../../../src";

describe("github issues > #112 Circular dependencies inside a scoped container", function() {

    beforeEach(() => Container.reset());

    it("should work properly", function() {

        @Service()
        class A {
            @Inject(() => B) public b: any;
        }

        @Service()
        class B {
            @Inject(() => A) public a: any;
        }

        const aFromGlobal = Container.get(A);
        const bFromGlobal = Container.get(B);

        expect(aFromGlobal).toEqual(bFromGlobal.a);
        expect(bFromGlobal).toEqual(aFromGlobal.b);

        const scopedContainer = Container.of("foo");

        const aFromScoped = scopedContainer.get(A);
        const bFromScoped = scopedContainer.get(B);

        expect(aFromScoped).toEqual(bFromScoped.a);
        expect(bFromScoped).toEqual(aFromScoped.b);

        expect(aFromScoped).not.toEqual(aFromGlobal);
        expect(bFromScoped).not.toEqual(bFromGlobal);
    });

});
```

**Comment by Dzeri96** [comment]
Can we revive this PR? I'm having to work around this issue somehow

**Comment by NoNameProvided** [comment]
Fixed in bf973acb5759ea413988a335b1fe56ec71a06c6d. Will be included in the next release.

**Comment by NoNameProvided** [comment]
This has been released as `0.9.0`.
